src/Avalonia.Visuals/Rendering/DirtyVisuals.cs

```diff
@@ -89,19 +100,69 @@ public bool Remove(IVisual visual)
         /// <returns>A collection of visuals.</returns>
         public IEnumerator<IVisual> GetEnumerator()
         {
-            foreach (var i in _inner)
+            using (DeferChanges())
             {
-                foreach (var j in i.Value)
+                foreach (var i in _inner)
                 {
-                    yield return j;
+                    foreach (var j in i.Value)
+                    {
+                        yield return j;
+                    }
                 }
             }
         }
 
+        private DeferDisposer DeferChanges()
+        {
+            _deferring = true;
+            return new DeferDisposer(this);
+        }
+
+        private void EndDefer()
+        {
+            if (!_deferring) return;
+
+            _deferring = false;
+
+            foreach (var change in _deferredChanges)
+            {
+                switch (change.change)
+                {
+                    case DeferredChange.Add:
+                        Add(change.visual);
+                        break;
+                    case DeferredChange.Remove:
+                        Remove(change.visual);
+                        break;
+                    case DeferredChange.Clear:
+                        Clear();
+                        break;
+                }
+            }
+
+            _deferredChanges.Clear();
+        }
+
         /// <summary>
         /// Gets the dirty visuals, in ascending order of distance to their root.
         /// </summary>
         /// <returns>A collection of visuals.</returns>
         IEnumerator IEnumerable.GetEnumerator() => GetEnumerator();
+
+        private struct DeferDisposer : IDisposable
```

Comment:
Do we really need to use the `using` pattern for this? Setting `_deferring` at the beginning and calling `EndDefer` in a `finally` block would work just as well for me. Feels like it'd be more performant too...

Reply:
`using` is the same as `try..finally`. There's no perf issue since this is a struct. But it's a bit redundant, you're right. I had it first because I made the method `DeferChanges` public, but then realized it's only ever needed during enumeration.